File: floddFill.js
/**
// Time Complexity : O(M*N)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No 




* @param {number[][]} image
 * @param {number} sr
 * @param {number} sc
 * @param {number} newColor
 * @return {number[][]}
 */

var floodFill = function(image, sr, sc, newColor) {
    if(image == null || image.length == 0) return image
    
      
    let ogColor = image[sr][sc];
      if(ogColor == newColor) return image
  
    dfs(image, sr, sc, newColor)
    return image
  
    function dfs(image, i, j, newColor){
        //base
      if(i < 0 || i >= image.length || j < 0 || j >= image[0].length || image[i][j] != ogColor) return 
        //logic
      let dirs = [[0,1],[0,-1],[-1,0],[1,0]];
      image[i][j] = newColor;
      for(let dir of dirs){
          let r = dir[0] + i
          let c = dir[1] + j
  
          dfs(image, r, c, newColor)
      }
   
    }
  };

File: floddfill.java
import java.util.LinkedList;
import java.util.List;

// Time Complexity : O(nm)
// Space Complexity :O(d)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : NO


// Your code here along with comments explaining your approach
class Solution {
    public int[][] floodFill(int[][] image, int sr, int sc, int newColor) {
        //edge case
        if(image == null || image.length==0) return image;
        int oldColor = image[sr][sc];
        if(oldColor==newColor) return image;
        color(image,sr,sc,oldColor,newColor);

        return image;
    }
    private void color(int[][] image, int sr, int sc,int oldColor, int newColor)
    {
        if(image[sr][sc]==oldColor)
        {
            image[sr][sc] = newColor;
            for(int[] p:neighbour(image,sr,sc))
            {
                int a= p[0];
                int b = p[1];
               color(image,a,b,oldColor,newColor);
            }
        }
    }
    private List<int[]> neighbour(int[][] image , int sr, int sc)
    {
        List<int []> result = new LinkedList<>();
        int[][] dirs={{0,1},{0,-1},{1,0},{-1,0}};
        for(int[] dir:dirs)
        {
            int R = sr+dir[0];
            int C = sc+dir[1];
            if(R>=0 && R < image.length && C >=0 && C <image[0].length)
            {
                result.add(new int[]{R,C});
            }
        }
        return result;
    }
}